File: packages/python/src/daraja/utils/token_cache.py
from __future__ import annotations

import threading
import time
from typing import TYPE_CHECKING, Protocol

if TYPE_CHECKING:
    import redis


class SharedTokenCache(Protocol):
    def get(self, key: str) -> str | None: ...
    def set(self, key: str, token: str, ttl_sec: int) -> None: ...


class InMemorySharedTokenCache:
    def __init__(self) -> None:
        self._cache: dict[str, tuple[str, float]] = {}
        self._lock = threading.Lock()
        self._timer: threading.Timer | None = None
        self._start_cleanup()

    def get(self, key: str) -> str | None:
        with self._lock:
            entry = self._cache.get(key)
            if entry is None:
                return None
            token, expires_at = entry
            if time.time() >= expires_at:
                del self._cache[key]
                return None
            return token

    def set(self, key: str, token: str, ttl_sec: int) -> None:
        with self._lock:
            self._cache[key] = (token, time.time() + ttl_sec)

    def _start_cleanup(self) -> None:
        self._cleanup()
        self._timer = threading.Timer(60.0, self._start_cleanup)
        self._timer.daemon = True
        self._timer.start()

    def _cleanup(self) -> None:
        now = time.time()
        with self._lock:
            expired = [k for k, (_, exp) in self._cache.items() if now >= exp]
            for k in expired:
                del self._cache[k]

    def dispose(self) -> None:
        if self._timer:
            self._timer.cancel()
        with self._lock:
            self._cache.clear()


class RedisTokenCache:
    def __init__(self, url: str) -> None:
        self._url = url
        self._client: redis.Redis | None = None  # type: ignore
        self._connect()

    def _connect(self) -> None:
        try:
            import redis
            self._client = redis.Redis.from_url(
                self._url,
                decode_responses=True,
                socket_connect_timeout=5,
            )
            self._client.ping()
        except Exception:
            self._client = None

    def get(self, key: str) -> str | None:
        if self._client is None:
            return None
        try:
            val = self._client.get(key)
            return val if val is not None else None
        except Exception:
            return None

    def set(self, key: str, token: str, ttl_sec: int) -> None:
        if self._client is None:
            return
        try:
            self._client.setex(key, ttl_sec, token)
        except Exception:
            pass

    def close(self) -> None:
        if self._client is not None:
            try:
                self._client.close()
            except Exception:
                pass


def build_token_cache_key(consumer_key: str) -> str:
    return f"mpesa:token:{consumer_key}"

File: sdks/typescript/src/utils/token-cache.ts
export interface SharedTokenCache {
  get(key: string): Promise<string | null>;
  set(key: string, token: string, ttlSec: number): Promise<void>;
}

export class InMemorySharedTokenCache implements SharedTokenCache {
  private cache = new Map<string, { token: string; expiresAt: number }>();
  private cleanupTimer: ReturnType<typeof setInterval>;

  constructor() {
    this.cleanupTimer = setInterval(() => this.cleanup(), 60_000);
  }

  async get(key: string): Promise<string | null> {
    const entry = this.cache.get(key);
    if (!entry) return null;
    if (Date.now() >= entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }
    return entry.token;
  }

  async set(key: string, token: string, ttlSec: number): Promise<void> {
    this.cache.set(key, { token, expiresAt: Date.now() + ttlSec * 1000 });
  }

  private cleanup(): void {
    const now = Date.now();
    for (const [key, entry] of this.cache) {
      if (now >= entry.expiresAt) this.cache.delete(key);
    }
  }

  dispose(): void {
    clearInterval(this.cleanupTimer);
    this.cache.clear();
  }
}

export class RedisTokenCache implements SharedTokenCache {
  private client: import("redis").RedisClientType | null = null;
  private clientPromise: Promise<import("redis").RedisClientType> | null = null;

  constructor(url: string) {
    this.clientPromise = this.init(url);
  }

  private async init(url: string): Promise<import("redis").RedisClientType> {
    try {
      const { createClient } = await import("redis");
      const client = createClient({ url }) as import("redis").RedisClientType;
      await client.connect();
      this.client = client;
      return client;
    } catch {
      return null as unknown as import("redis").RedisClientType;
    }
  }

  async get(key: string): Promise<string | null> {
    if (!this.client) await this.clientPromise;
    if (!this.client) return null;
    try {
      return await this.client.get(key);
    } catch {
      return null;
    }
  }

  async set(key: string, token: string, ttlSec: number): Promise<void> {
    if (!this.client) await this.clientPromise;
    if (!this.client) return;
    try {
      await this.client.set(key, token, { EX: ttlSec });
    } catch { /* ignore */ }
  }

  async disconnect(): Promise<void> {
    if (this.client) {
      try { await this.client.quit(); } catch { /* ignore */ }
    }
  }
}

export function buildTokenCacheKey(consumerKey: string): string {
  return `mpesa:token:${consumerKey}`;
}
